File: test/java/src/test/java/id/dana/util/BNIHashUtil.java
package id.dana.util;

import java.util.Base64;

public final class BNIHashUtil {

  private BNIHashUtil() {}

  public static String hashData(String jsonData, String clientId, String secretKey) {
    String time = String.valueOf(System.currentTimeMillis());
    time = time.substring(0, Math.min(time.length(), 10));
    time = new StringBuilder(time).reverse().toString();
    return doubleEncrypt(time + "." + jsonData, clientId, secretKey);
  }

  private static String doubleEncrypt(String input, String clientId, String secretKey) {
    byte[] encrypted = encrypt(input.getBytes(), clientId);
    encrypted = encrypt(encrypted, secretKey);
    return Base64.getEncoder().encodeToString(encrypted)
        .replaceAll("=+$", "")
        .replace('+', '-')
        .replace('/', '_');
  }

  private static byte[] encrypt(byte[] data, String key) {
    byte[] result = new byte[data.length];
    int keyLen = key.length();
    for (int i = 0; i < data.length; i++) {
      char keyChar = key.charAt((i + keyLen - 1) % keyLen);
      result[i] = (byte) ((data[i] + keyChar) % 128);
    }
    return result;
  }
}

File: test/php/helper/BniHashUtil.php
<?php

namespace DanaUat\Helper;

final class BniHashUtil
{
    private function __construct()
    {
    }

    public static function hashData(string $jsonData, string $clientId, string $secretKey): string
    {
        $time = substr((string) (int) (microtime(true) * 1000), 0, 10);
        $time = strrev($time);
        return self::doubleEncrypt($time . '.' . $jsonData, $clientId, $secretKey);
    }

    private static function doubleEncrypt(string $input, string $clientId, string $secretKey): string
    {
        $encrypted = self::encrypt($input, $clientId);
        $encrypted = self::encrypt($encrypted, $secretKey);
        return rtrim(strtr(base64_encode($encrypted), '+/', '-_'), '=');
    }

    private static function encrypt(string $data, string $key): string
    {
        $keyLen = strlen($key);
        $result = '';
        for ($i = 0, $len = strlen($data); $i < $len; $i++) {
            $keyChar = $key[($i + $keyLen - 1) % $keyLen];
            $result .= chr((ord($data[$i]) + ord($keyChar)) % 128);
        }
        return $result;
    }
}
